pkg/instrumentation/auto/callback.go

```diff
@@ -0,0 +1,54 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package auto
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type objectCallbackFunc func(client.Object) bool
+
+func chainCallbacks(fns ...objectCallbackFunc) objectCallbackFunc {
+	return func(obj client.Object) bool {
+		for _, fn := range fns {
+			if !fn(obj) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func (m *AnnotationMutators) patchFunc(ctx context.Context, callback objectCallbackFunc) objectCallbackFunc {
+	return func(obj client.Object) bool {
+		patch := client.StrategicMergeFrom(obj.DeepCopyObject().(client.Object))
```

Comment:
Need the deep copy because the object is passed around as a reference. The other option is to create our own `client.Patch` type that takes in the marshalled JSON. That might be better.